### DISCUSSION TITLE
Start only one rust-analyzer in different workspace root files

### DISCUSSION THREAD

**Discussion by xzbdmw** [mrcjkb/rustaceanvim#240] [Q&A] [answered]
### Neovim version (nvim -v)

0.10

### Operating system/version

macos14.2

### Output of :checkhealth rustaceanvim

```console
==============================================================================
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- WARNING dap not installed. Needed for debugging features [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap)

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 1.76.0 (07dca489 2024-02-04)
- OK Cargo: found cargo 1.76.0 (c84b36747 2024-01-18)
- OK rustc: found rustc 1.76.0 (07dca489a 2024-02-04)
- OK debug adapter: found codelldb 

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

### How to reproduce the issue

```console
open two rust file at different folders.
```

### Expected behaviour

start two rust-analyzer instance as lspconfig do/

### Actual behaviour

the two different project share signle server. This makes nvimtree sync with cwd broken, make session manage plugin broken.
Why does the nvimtreetoggle command keep switching between different cwds? Because lsp modified cwd, when I switched to the second workspace, lsp still ran in the workspace of the first file, as can be seen from Lspinfo.

rustacean: start only one server, cause cwd to change when toggle tree.
https://github.com/mrcjkb/rustaceanvim/assets/97848247/b4240ee6-19c5-4926-bee5-3e160de65e71

mason+lspconfig: start two server, file tree woks fine.
https://github.com/mrcjkb/rustaceanvim/assets/97848247/1515726b-f40e-493f-b28c-6852c18f6e9b

### The minimal config used to reproduce this issue.

```Lua
default config
```

**Comment by mrcjkb** [comment]
Hey 👋

Starting a single client and registering workspace changes via [`workspace/didChangeWorkspaceFolders`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWorkspaceFolders) is the intended behaviour.

This sounds like a bug in nvim-tree.

Converting to a discussion.

**Reply by xzbdmw** [reply]
But lspconfig handles this differently，other server that not active is running in single file mode and don’t cause these issues

**Reply by mrcjkb** [reply]
Oh, you mean when opening standalone files that are not part of a Cargo project?
That's a rust-analyzer issue.

See :

- https://github.com/mrcjkb/rustaceanvim/issues/190

and

- https://github.com/rust-lang/rust-analyzer/issues/14318

The way lspconfig handles it is not something I want to implement for this plugin, as it can potentially lead to other issues (e.g. spawning too many processes when navigating outside of a project).

**Reply by xzbdmw** [reply]
What I mean is that when I open a file in another workspace, the inactive lspserver runs in single file mode, If I switch back to the original file at this point, the inactive file will return to the mode with a workspace.
<img width="880" alt="截屏2024-02-20 03 15 44" src="https://github.com/mrcjkb/rustaceanvim/assets/97848247/3331ca77-d195-46a4-b45e-4fbac3b2b3ad">
The third one is the active file before I switch and now become inactive thus running in single file mode

**Reply by mrcjkb** [reply, marked as answer]
I see. Again, this is the intended behaviour. See also [this comment on a related lspconfig issue](https://github.com/neovim/nvim-lspconfig/issues/842#issuecomment-1172890154).

Please keep in mind that Neovim has no concept of a workspace. That is something that is implemented by plugins like [`project.nvim`](https://github.com/ahmedkhalf/project.nvim).
If nvim-tree (or some project management plugin you're using) is using the LSP client's `config.root_dir` as the source of truth to set the `cwd` or for the current buffer's project root, and is ignoring the `workspace_folders`, that's a bug.

It should be using the client's `workspace_folders` property, comparing the `uri`s with the current buffer's path, and then fall back to the `config.root_dir` if it can't find a match in the `workspace_folders`.

If I have two files from different Cargo projects open, these are the `workspace_folders`:

```lua
> :lua =vim.lsp.get_clients()[1].workspace_folders
{ {
    name = "/home/mrcjk/git/github/forks/nvim/rocks",
    uri = "file:///home/mrcjk/git/github/forks/nvim/rocks"
}, {
    name = "/home/mrcjk/git/github/forks/tuxedo-rs",
    uri = "file:///home/mrcjk/git/github/forks/tuxedo-rs"
} }
```

It looks like project.nvim [uses `root_dir` as the source of truth](https://github.com/ahmedkhalf/project.nvim/blob/8c6bad7d22eef1b71144b401c9f74ed01526a4fb/lua/project_nvim/project.lua#L25).

**Reply by xzbdmw** [reply]
Thank you for your patient answer. I disabled project.nvim and everything is fine, Do you have any other suggestions for similar plugins:)

**Reply by mrcjkb** [reply]
No problem :smile: 

I have project.nvim installed myself, so I opened [a PR with a fix](https://github.com/ahmedkhalf/project.nvim/pull/147).
It looks like the plugin hasn't been touched in about a year though, so I don't have high hopes about it getting merged anytime soon.

Personally, I haven't really needed it, because I mostly use tmux (and [tmux-sessionizer](https://github.com/jrmoulton/tmux-sessionizer)) to switch between projects.
That's why I never noticed the bug :)
